psi4/src/psi4/libfock/v.cc

```diff
@@ -1289,53 +1517,140 @@ SharedMatrix RV::compute_hessian() {
-            int A = primary_->function_to_center(function_map[ml]);
-            Hp[3 * A + 1][3 * A + 0] += C_DDOT(npoints, &pTx2[0][ml], max_functions, &Up[0][ml], max_functions);
-            Hp[3 * A + 1][3 * A + 1] += C_DDOT(npoints, &pTy2[0][ml], max_functions, &Up[0][ml], max_functions);
-            Hp[3 * A + 1][3 * A + 2] += C_DDOT(npoints, &pTz2[0][ml], max_functions, &Up[0][ml], max_functions);
+            for (int nl = 0; nl < nlocal; nl++) {
+                double D = Dp[ml][nl];
+                pHXY[ml][nl] += pTx2[ml][nl] * D;
+                pHYY[ml][nl] += pTy2[ml][nl] * D;
+                pHZY[ml][nl] += pTz2[ml][nl] * D;
+            }
         }
-        // x derivatives
+
+        // z derivatives
         for (int P = 0; P < npoints; P++) {
             std::fill(Up[P], Up[P] + nlocal, 0.0);
-            C_DAXPY(nlocal, 2.0 * w[P] * v_rho_a[P], pTz2[P], 1, Up[P], 1);
+            if (std::fabs(rho_a[P]) > v2_rho_cutoff_) {
+                C_DAXPY(nlocal, 2.0 * w[P] * v_rho_a[P], phi_z[P], 1, Up[P], 1);
+            }
         }
+        C_DGEMM('T', 'N', nlocal, nlocal, npoints, 1.0, phi_x[0], coll_funcs, Up[0], max_functions, 0.0, pTx2[0], max_functions);
+        C_DGEMM('T', 'N', nlocal, nlocal, npoints, 1.0, phi_y[0], coll_funcs, Up[0], max_functions, 0.0, pTy2[0], max_functions);
+        C_DGEMM('T', 'N', nlocal, nlocal, npoints, 1.0, phi_z[0], coll_funcs, Up[0], max_functions, 0.0, pTz2[0], max_functions);
+        for (int ml = 0; ml < nlocal; ml++) {
+            for (int nl = 0; nl < nlocal; nl++) {
+                double D = Dp[ml][nl];
+                pHXZ[ml][nl] += pTx2[ml][nl] * D;
+                pHYZ[ml][nl] += pTy2[ml][nl] * D;
+                pHZZ[ml][nl] += pTz2[ml][nl] * D;
+            }
+        }
+
+        // Accumulate contributions to the full Hessian
         for (int ml = 0; ml < nlocal; ml++) {
             int A = primary_->function_to_center(function_map[ml]);
-            Hp[3 * A + 2][3 * A + 0] += C_DDOT(npoints, &pTx2[0][ml], max_functions, &Up[0][ml], max_functions);
-            Hp[3 * A + 2][3 * A + 1] += C_DDOT(npoints, &pTy2[0][ml], max_functions, &Up[0][ml], max_functions);
-            Hp[3 * A + 2][3 * A + 2] += C_DDOT(npoints, &pTz2[0][ml], max_functions, &Up[0][ml], max_functions);
+            for (int nl = 0; nl < nlocal; nl++) {
+                int B = primary_->function_to_center(function_map[nl]);
+                Hp[3 * A + 0][3 * B + 0] += pHXX[ml][nl];
+                Hp[3 * A + 1][3 * B + 0] += pHYX[ml][nl];
+                Hp[3 * A + 2][3 * B + 0] += pHZX[ml][nl];
+                Hp[3 * A + 0][3 * B + 1] += pHXY[ml][nl];
+                Hp[3 * A + 1][3 * B + 1] += pHYY[ml][nl];
+                Hp[3 * A + 2][3 * B + 1] += pHZY[ml][nl];
+                Hp[3 * A + 0][3 * B + 2] += pHXZ[ml][nl];
+                Hp[3 * A + 1][3 * B + 2] += pHYZ[ml][nl];
+                Hp[3 * A + 2][3 * B + 2] += pHZZ[ml][nl];
```

Comment:
Can we add a small blurb that this is *not* symmetric? Not sure why, but that caught me for a bit.